File: Sources/ClipDeckKit/Variables/TemplateRenderer.swift
import Foundation

/// Uma variável encontrada no corpo de um template.
public struct TemplateVariable: Equatable, Identifiable {
    public let name: String
    /// Built-ins são resolvidas sozinhas e não viram campo no formulário.
    public let isBuiltIn: Bool

    public init(name: String, isBuiltIn: Bool) {
        self.name = name
        self.isBuiltIn = isBuiltIn
    }

    public var id: String { name }

    /// Rótulo legível: "proximo_passo" vira "Próximo passo"? Não — trocar
    /// underscore por espaço e capitalizar a primeira letra basta e não inventa
    /// acentuação que o usuário não escreveu.
    public var label: String {
        let spaced = name.replacingOccurrences(of: "_", with: " ")
        return spaced.prefix(1).uppercased() + spaced.dropFirst()
    }
}

/// Interpreta `{{variáveis}}` no corpo dos templates.
public enum TemplateRenderer {
    /// Variáveis resolvidas automaticamente, sem perguntar nada ao usuário.
    public static let builtInNames: Set<String> = ["data", "hora", "clipboard"]

    /// Extrai as variáveis na ordem em que aparecem, sem repetir.
    ///
    /// A ordem importa: é a ordem dos campos no formulário, e ela precisa bater
    /// com a leitura do texto para o preenchimento fazer sentido.
    public static func variables(in body: String) -> [TemplateVariable] {
        var found: [TemplateVariable] = []
        var seen: Set<String> = []

        for name in rawNames(in: body) {
            guard !seen.contains(name) else { continue }
            seen.insert(name)
            found.append(
                TemplateVariable(name: name, isBuiltIn: builtInNames.contains(name))
            )
        }
        return found
    }

    /// Só as variáveis que precisam ser preenchidas à mão.
    public static func fillableVariables(in body: String) -> [TemplateVariable] {
        variables(in: body).filter { !$0.isBuiltIn }
    }

    /// Substitui as variáveis pelos valores.
    ///
    /// Variáveis sem valor informado ficam como estão, em vez de virarem string
    /// vazia: um buraco visível no texto é melhor que uma frase que perdeu uma
    /// palavra sem você perceber.
    public static func render(
        _ body: String,
        values: [String: String],
        clipboard: String? = nil,
        now: Date = Date()
    ) -> String {
        var output = ""
        var remainder = Substring(body)

        while let open = remainder.range(of: "{{") {
            output += remainder[remainder.startIndex..<open.lowerBound]
            let afterOpen = remainder[open.upperBound...]

            guard let close = afterOpen.range(of: "}}") else {
                // Sem fechamento: o resto é texto literal, incluindo o "{{".
                output += remainder[open.lowerBound...]
                return output
            }

            let name = afterOpen[afterOpen.startIndex..<close.lowerBound]
                .trimmingCharacters(in: .whitespaces)

            if let resolved = resolve(name: name, values: values, clipboard: clipboard, now: now) {
                output += resolved
            } else {
                output += "{{\(name)}}"
            }

            remainder = afterOpen[close.upperBound...]
        }

        output += remainder
        return output
    }

    private static func resolve(
        name: String,
        values: [String: String],
        clipboard: String?,
        now: Date
    ) -> String? {
        // Valor informado tem prioridade sobre built-in: se o template usa
        // {{data}} e o usuário digitou uma data, vale a dele.
        if let provided = values[name], !provided.isEmpty { return provided }

        switch name {
        case "data": return formatted(now, style: "dd/MM/yyyy")
        case "hora": return formatted(now, style: "HH:mm")
        case "clipboard": return clipboard
        default: return nil
        }
    }

    private static func formatted(_ date: Date, style: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "pt_BR")
        formatter.dateFormat = style
        return formatter.string(from: date)
    }

    /// Nomes na ordem de aparição, com repetições.
    private static func rawNames(in body: String) -> [String] {
        var names: [String] = []
        var remainder = Substring(body)

        while let open = remainder.range(of: "{{") {
            let afterOpen = remainder[open.upperBound...]
            guard let close = afterOpen.range(of: "}}") else { break }

            let name = afterOpen[afterOpen.startIndex..<close.lowerBound]
                .trimmingCharacters(in: .whitespaces)
            if !name.isEmpty { names.append(name) }

            remainder = afterOpen[close.upperBound...]
        }
        return names
    }
}

File: windows/src/ClipDeck.Core/Variables/TemplateRenderer.cs
using System.Globalization;
using System.Text;

namespace ClipDeck.Core.Variables;

/// <summary>Uma variável encontrada no corpo de um template.</summary>
public readonly record struct TemplateVariable(string Name, bool IsBuiltIn)
{
    /// <summary>
    /// Rótulo legível. Trocar underscore por espaço e subir a primeira letra
    /// basta — inventar acentuação que o usuário não escreveu seria pior.
    /// </summary>
    public string Label
    {
        get
        {
            var spaced = Name.Replace('_', ' ');
            return spaced.Length == 0 ? spaced : char.ToUpperInvariant(spaced[0]) + spaced[1..];
        }
    }
}

/// <summary>Interpreta {{variáveis}} no corpo dos templates.</summary>
public static class TemplateRenderer
{
    /// <summary>Variáveis resolvidas sozinhas, sem perguntar nada ao usuário.</summary>
    public static readonly IReadOnlySet<string> BuiltInNames =
        new HashSet<string> { "data", "hora", "clipboard" };

    private static readonly CultureInfo Culture = new("pt-BR");

    /// <summary>
    /// Extrai as variáveis na ordem em que aparecem, sem repetir. A ordem importa:
    /// é a ordem dos campos no formulário.
    /// </summary>
    public static List<TemplateVariable> Variables(string body)
    {
        var found = new List<TemplateVariable>();
        var seen = new HashSet<string>(StringComparer.Ordinal);

        foreach (var name in RawNames(body))
        {
            if (!seen.Add(name)) continue;
            found.Add(new TemplateVariable(name, BuiltInNames.Contains(name)));
        }

        return found;
    }

    /// <summary>Só as variáveis que precisam ser preenchidas à mão.</summary>
    public static List<TemplateVariable> FillableVariables(string body) =>
        Variables(body).Where(v => !v.IsBuiltIn).ToList();

    /// <summary>
    /// Substitui as variáveis pelos valores. Variáveis sem valor informado ficam
    /// como estão, em vez de virarem string vazia: um buraco visível no texto é
    /// melhor que uma frase que perdeu uma palavra sem você perceber.
    /// </summary>
    public static string Render(
        string body,
        IReadOnlyDictionary<string, string> values,
        string? clipboard = null,
        DateTimeOffset? now = null)
    {
        var moment = now ?? DateTimeOffset.Now;
        var output = new StringBuilder(body.Length);
        var index = 0;

        while (index < body.Length)
        {
            var open = body.IndexOf("{{", index, StringComparison.Ordinal);
            if (open < 0)
            {
                output.Append(body, index, body.Length - index);
                break;
            }

            output.Append(body, index, open - index);

            var close = body.IndexOf("}}", open + 2, StringComparison.Ordinal);
            if (close < 0)
            {
                // Sem fechamento: o resto é texto literal, incluindo o "{{".
                output.Append(body, open, body.Length - open);
                break;
            }

            var name = body[(open + 2)..close].Trim();
            var resolved = Resolve(name, values, clipboard, moment);
            output.Append(resolved ?? $"{{{{{name}}}}}");

            index = close + 2;
        }

        return output.ToString();
    }

    private static string? Resolve(
        string name,
        IReadOnlyDictionary<string, string> values,
        string? clipboard,
        DateTimeOffset now)
    {
        // Valor informado tem prioridade sobre built-in: se o template usa
        // {{data}} e o usuário digitou uma data, vale a dele.
        if (values.TryGetValue(name, out var provided) && provided.Length > 0) return provided;

        return name switch
        {
            "data" => now.ToString("dd/MM/yyyy", Culture),
            "hora" => now.ToString("HH:mm", Culture),
            "clipboard" => clipboard,
            _ => null,
        };
    }

    /// <summary>Nomes na ordem de aparição, com repetições.</summary>
    private static IEnumerable<string> RawNames(string body)
    {
        var index = 0;
        while (index < body.Length)
        {
            var open = body.IndexOf("{{", index, StringComparison.Ordinal);
            if (open < 0) yield break;

            var close = body.IndexOf("}}", open + 2, StringComparison.Ordinal);
            if (close < 0) yield break;

            var name = body[(open + 2)..close].Trim();
            if (name.Length > 0) yield return name;

            index = close + 2;
        }
    }
}
